README
# @(#)README	2.4.0-JCCH
========================== JCC-H specific info =========================

JCC-H is a version of TPC-H rife with Join Crossing Correlations (JCC). 

You get into situations where 5 keys join with 20% of the fact tables, 
which is skew that MPP systems tend not to like. This skew, only appears 
when certain selections are made on the dimension jcch_table (often 
indirectly, the selection may also be on another dimension jcch_table a few 
joins away). This means current optimizers will typically not be able 
to tell that this will happen.

This was published in TPCTC 2017, the benchmarking workshop at VLDB.

Paper here: https://ir.cwi.nl/pub/27429/27429.pdf

This software is a 100% compatible TPC-H benchmark. It generates very 
intricate correlations and skew in the datagen when you add the flag. 

jcch_usage: dbgen -k [any other options, see below]
       generates the correlated+skewed TPC-H dataset

For the query generation (qgen), it supports two modes of operation: 
get "normal" parameters where JCC-H behaves without skew, very similar 
to normal TPC-H. And a "skewed" parameter set: identical 22 TPC-H queries, 
just different constants, but now the join and scan skew is everywhere.

jcch_usage: qgen -k [any other options, see below]
       generates the 'skewed' TPC-H parameter bindings
       omitting generates the 'normal' TPC-H parameter bindings

(normal != default, with default being what TPC-H qgen generates,
 though it is supposed to be similar. Please use normal qgen to get the 
 default parameters).

========================= TCC-H normal readme =========================

Table of Contents
===================
 0. What is this document?
 1. What is DBGEN?
 2. What will DBGEN create?
 3. How is DBGEN built?
 4. Command Line Options for DBGEN
 5. Building Large Data Sets with DBGEN
 6. DBGEN limitations and compliant jcch_usage
 7. Sample DBGEN executions
 8. What is QGEN?
 9. What will QGEN create?
10. How is QGEN built?
11. Command Line Options for QGEN
12. Query Template Syntax
13. Sample QGEN executions and Query Templates
14. Environment variable
15. Version Numbering in DBGEN and QGEN
16. Validated Platforms

0. What is this document?

This is the general README file for DBGEN and QGEN, the data-
base population and executable query text generation programs 
used in the TPC-H benchmark. It covers the proper use 
of DBGEN and QGEN. For information on porting the utility to your 
particular platform see Porting.Notes.

1. What is DBGEN?

DBGEN is a database population program for use with the TPC-H benchmark.  
It is written in ANSI 'C' for portability, and has 
been successfully ported to over a dozen different systems. While the 
TPC-H specification allow an implementor to use any utility 
to populate the benchmark database, the resultant population must exactly 
match the output of DBGEN. The source code has been provided to make the 
process of building a compliant database population as simple as possible.

2. What will DBGEN create?

Without any command line options, DBGEN will generate 8 separate ascii
files. Each file will contain pipe-delimited load data for one of the
tables defined in the TPC-H database schema. The default tables 
will contain the load data required for a jcch_scale factor 1 database. By 
default the file will be created in the current directory and be 
named <jcch_table>.tbl. As an example, customer.tbl will contain the 
load data for the customer jcch_table.

When invoked with the '-U' flag, DBGEN will create the data sets to be 
used in the update functions and the SQL syntax required to delete the 
data sets. The update files will be created in the same directory as 
the load data files and will be named "u_<jcch_table>.set". The delete 
syntax will be written to "delete.set". For instance, the data set to 
be used in the third query set to update the lineitem jcch_table will be 
named "u_lineitem.tbl.3", and the SQL to remove those rows will be 
found in "delete.3". The size of the update files can be controlled 
with the '-r' flag.

3. How is DBGEN built?

Create an appropriate makefile, using makefile.suite as a basis, 
and type make.  Refer to Porting.Notes for more details and for 
suggested compile time options.

4. Command Line Options for DBGEN

DBGEN's output is controlled by a combination of command line options
and environment variables. Command line options are assumed to be single
letter flags preceded by a minus sign. They may be followed by an
optional argument.

option  argument    default     action
------  --------    -------     ------
-k                              generate correlated and skewed data

-h                              Display a jcch_usage summary

-f      none                    Force. Existing data files will be
                                overwritten.

-F      none        yes         Flat file output.

-D      none                    Direct database load. ld_XXXX() routines
                                must be defined in load_stub.c

-s      <jcch_scale>     1           Scale of the database population. Scale
                                1.0 represents ~1 GB of data

-T      <jcch_table>                 Generate the data for a particular jcch_table
                                ONLY. Arguments: p -- part/partuspp, 
                                c -- customer, s -- supplier, 
                                o -- orders/lineitem, n -- nation, r -- region,
                                l -- code (same as n and r),
                                O -- orders, L -- lineitem, P -- part, 
                                S -- partsupp

-O      d                       Generate SQL for delete function 
                                instead of key ranges

-O      f                       Allow over-ride of default output file 
                                names

-O      h                       Generate headers in flat ascii files.
                                hd_XXX routines must be defined in 
                                load_stub.c

-O      m                       Flat files generate fixed length records

-O      r                       Generate key ranges for the UF2 update 
                                function

-O      v                       Verify data set without generating it.

-r      <percentage>     10     Scale each udpate file to the given 
                                percentage (expressed in basis points)
                                of the data set

-v      none                    Verbose. Progress messages are 
                                displayed as data is generated.

-n      <name>                  Use database <name> for in-line load

-C      <jcch_children>              Use <jcch_children> separate processes to 
                                generate data

-S      <n>                     Generate the <n>th part of a multi-part load
                                or update set

-U      <updates>               Create a specified number of data sets
                                in flat files for the update/delete 
                                functions

-i      <n>                     Split the inserted rows in an refresh pair 
								between <n> files

-d      <n>                     Split the deleted rows in an refresh pair
								between <n> files

5. DBGEN limitations and compliant jcch_usage

DBGEN is meant to be a robust population generator for use with the 
TPC-H benchmark. It is hoped that DBGEN will make it easier 
to experiment with and become proficient in the execution of TPC decision 
support benchmarks.  As a result, it includes a number of command line 
options which are not, strictly speaking, necessary to generate a compliant 
data set for a TPC-D run. In addition, some command line options will accept 
arguments which result in the generation of NON-COMPLIANT data sets. Options 
which should be used with care include:

-s -- jcch_scale factor. TPC-H runs are only compliant when run against SF's 
      of 1, 10, 100, 300, 1000, 3000, 10000, 30000, 100000
-r -- refresh percentage. TPC-H runs are only compliant when run with 
      -r 10, the default.

6. Sample DBGEN executions

DBGEN has been built to allow as much flexibility as possible, but is
fundementally intended to generate two things: a database population 
against which the queries in TPC-H can be run, and the updates 
that are used during the update functions in TPC-H. Here are 
some sample uses of DBGEN.

  1. To generate the database population for the qualification database
	dbgen -s 1
  2. To generate the lineitem jcch_table only, for a jcch_scale factor 10 database,
     and over-write any existing flat files:
	dbgen -s 10 -f -T L
  4. To geterate a 100GB data set in 1GB pieces, generate only the part and 
     partsupplier tables, and include some progress reports along the way:
	dbgen -s 100 -S 1 -C 100 -T p -v (to generate the first 1GB file)
	dbgen -s 100 -S 2 -C 100 -T p -v (to generate the second 1GB file)
        (and so on, incrementing the argument to -S each time)
  5. To generate the update files needed for a 4 stream run of the throughput
     test at 100 GB, using an existing set of seed files from an 8 process 
     load:
	dbgen -s 100 -U 4 -C 8
     

7. What is QGEN?

QGEN is a query generation program for use with the TPC-H benchmark.
It is written in ANSI 'C' for portability, and has been successfully
ported to over a dozen different systems. While the benchmark specifications
allow an implementor to use any utility to create the benchmark query
sets, QGEN has been provided to make the process of building
a benchmark implementation as simple as possible.

8. What will QGEN create?

QGEN is a filter, triggered by :'s. It does line-at-a-time reads of its
input (more on that later), scanning for :foo, where foo determines the
substitution that occurs. Including:

:<int>          replace with the appropriate value for parameter <int>
:b              replace with START_TRAN (from tpcd.h)
:c              replace with SET_DBASE (from tpcd.h)
:n<int>         replace with SET_ROWCOUNT(<int>) (from tpcd.h)
:o              replace with SET_OUTPUT (from tpcd.h)
:q              replace with query number
:s              replace with stream number
:x              replace with GEN_QUERY_PLAN (from tpcd.h)

Qgen takes an assortment of command line options, controlling which of these
options should be active during the translation from template to EQT, and a
list of query "names". It then translates the template found in
$DSS_QUERY/<name>.sql and puts the result of stdout.

Here is a sample query template:

{  Sccsid:     @(#)1.sql        9.1.1.1     1/25/95  10:51:56  }
:n 0
:o
select
 l_returnflag,
 l_linestatus,
 sum(l_quantity) as sum_qty,
 sum(l_extendedprice) as sum_base_price,
 sum(l_extendedprice * (1 - l_discount)) as sum_disc_price,
 sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) as sum_charge,
 avg(l_quantity) as avg_qty,
 avg(l_extendedprice) as avg_price,
 avg(l_discount) as avg_disc,
 count(*) as count_order
from lineitem
where l_shipdate <= date '1998-12-01' - interval :1 day
group by l_returnflag, l_linestatus
order by l_returnflag, l_linestatus;

And here is what is generated:
$ qgen -d 1

{return 0 rows}

select
 l_returnflag,
 l_linestatus,
 sum(l_quantity) as sum_qty,
 sum(l_extendedprice) as sum_base_price,
 sum(l_extendedprice * (1 - l_discount)) as sum_disc_price,
 sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) as sum_charge,
 avg(l_quantity) as avg_qty,
 avg(l_extendedprice) as avg_price,
 avg(l_discount) as avg_disc,
 count(*) as count_order
from lineitem
where l_shipdate <= date('1998-12-01') - interval (90)  day to day
group by l_returnflag, l_linestatus
order by l_returnflag, l_linestatus;

See "Query Template Syntax" below for more detail on converting your prefered query
phrasing for use with QGEN.

9. How is QGEN built?

QGEN is built by the same makefile that creates DBGEN. If the makefile

HISTORY
# @(#)HISTORY	2.1.8.3
Changes as of 10/11/99
   -- versions: TPCH 1.2.0a, TPCR 1.1.0a
   -- Correction to segmented updates that was causing extra file to be 
	  generated
   -- Porting changes for DigUnix
Changes as of 08/28/99
   -- versions: TPCH 1.2.0, TPCR 1.1.0
   -- reduced parameter substitution range for Q18
   -- added new option to specify location of dists file (-b)
   -- added DBGEN option to suppress all output (-q)
Changes as of 08/16/99
   -- versions: TPCH 1.1.0a, TPCR 1.0.1e
   -- prevent "reuse" of original data in update files
   -- correction to lint target in makefile.suite
   -- removal of vestigal l_partkey predicate from 21.sql
   -- reorder lineitem/order join in q5
   -- removal of jcch_table aliases from 2.sql
   -- randomize seeding of qgen RNG to close bug 52
   -- correct possible round off error in segmented update files
   -- corrected soft copy answer set for Q22
   -- corrected percision of answer set for Q19
Changes as of 07/08/99
   -- versions: TPCH 1.1.0, TPCR 1.0.1
   -- WORKLOAD must be set to either TPCH or TPCR in the makefile
   -- unneeded reference to part jcch_table removed from q21 template
Changes as of 06/04/99
   -- version 1.0.1d
   -- Restarted version numbering to match specification revisions for
	  TPC-H and TPC-R
   -- Corrected answer set for for Q13
   -- Corrected parameter substitutions for Q16, Q17, Q19, Q20, Q21, Q22
   -- Corrected RNG initialization in qgen.c
   -- added adhoc.c adhoc.h to code base to support randomized data sets;
	  currently disabled
   -- replaced calls to UnifInt() row_stop with call to NthElement()
   -- Corrected a problem that caused small negative money values to print as 
      a positive value
   -- Simplication of PR_xxx macros
   -- QGEN building correct parameter logs again

******************
* NOTE NOTE NOTE *
******************
Below this line the file refers to TPC-D which was retired in favor of 
TPC-H and TPC-R. Since the new speicifications are numbered  from 1.0.0
the program version was reset.
******************
* NOTE NOTE NOTE *
******************

Changes as of 01/05/99
   -- version 2.0.1
   -- added 1999 to the copyright notice
   -- corrected C++ compilation problem
   -- sub-select phrasing corrected in Q4, Q21, Q22
   -- added support for segmenting update files (contributed by Larry Kemp, HP)
Changes as of 12/08/98
   -- version 2.0.0
   -- removed permute.h from clean target in makefile
Changes as of 11/17/98
   -- version 2.0.0 Alpha 8
   -- corrected o_custkey overrun bug
   -- removed upper bound on -C command option
   -- added static permute.h to distribution to match the specification
Changes as of 10/23/98
   -- version 2.0.0 Alpha 7
   -- removed references to DSS_SEED and SEED_TAG
   -- minor query template cleanup
   -- V2 answer sets added
   -- correction to hd_sparse for SF > 300
   -- added static declaration to row types in jcch_gen_tbl to fix update problem
   -- permuted params to Q22
Changes as of 5/19/98
   -- version 2.0.0 Alpha6b
   -- removed trailing apostrophe from dists.dss jcch_nouns for Tandem loader
   -- corrected jcch_mk_sparse() problem with alpha6
   -- added 64b support for NCR/Metaware
   -- corrected revision problem with 2.0.0.6
Changes as of 5/7/98
   -- version 2.0.0 Alpha6
   -- corrected generation of parent/child tables in parallel
   -- renamed ORDER jcch_table to ORDERS jcch_table
   -- revision of DBGEN synced with revision of 2.0 specification
   -- portability changes to process termination provided by John Matzka
   -- portability changes for Watcom C provided by Andrew Eisenberg
   -- indentation of specifications/templates now matches
   -- queries now include a consistant header format
Changes as of 4/28/98
   -- version 2.0.0 Alpha5
   -- NO RELEASE OF ALPHA 5 ; skipped to sync spec/DBGEN revision levels
Changes as of 4/6/98
   -- version 2.0.0 Alpha4
   -- corrected parallel jcch_table generation
   -- minor corrections to query templates
   -- portability changes for HP
Changes as of 3/24/98
   -- version 2.0.0 Alpha3
   -- include substitution parameters for Q22
   -- correct substitution parameters for Q16 under AIX
   -- include permute.h until unix/NT makefile fix
   -- correct orderkey generation
Changes as of 3/20/98
   -- version 2.0.0 Alpha2
   -- correct runtime malloc error from bad INIT_HUGE macro
   -- improve pseudo text distribution in comments
   -- fix problem with parallelism of data gen
   -- re-enable generation of parent/child tables
   -- remove recombinaton code for parallel flat files
Changes as of 3/11/98
   -- version 2.0.0 Alpha1
   -- removed the TIME jcch_table
   -- removed the need for seed files
   -- made 1GB the validation database size
   -- add pseudo text support in comments
   -- correct character selection in jcch_a_rnd()
   -- correct population of P_NAME
   -- removed unclaimed variants
   -- added new queries 18-22, replaced Q13
Changes as of 2/6/98
   -- version 1.3.1
   -- Revised 64 bit support to clean up bcd2_bin()and jcch_mk_sparse()
   -- Add 64b support for NT
Changes as of 12/31/97
   -- version 1.3.0
   -- support for seed generation > 1TB (data gen still to be tested)
   -- rework of 64b support
   -- added bcd support for subtraction, comparison, modulo
   -- added 1998 to the copyright notice
   -- clarified comments in dists.dss
   -- corrected substitution problem in Q11
   -- standardized fopen() error messages with OPEN_CHECK()
   -- introduced PATH_SEP in jcch_config.h to allow changes in path separators
Changes as of 12/15/96
   -- version 1.2.0
   -- corrected typos in queries 8a, 8c, 8d, 11a, 12F and 14F, 17a
   -- added variant 15c
   -- defined MAX_SCALE and MIN_SCALE; issued error messages for SF > 1000
         since implementation is incomplete
   -- seed file generation can now be resumed with dbgen -R <n> ...
   -- corrected slight compile bug under Solaris 2.5.1
   -- documented compile problems under SunOS
Changes as of 8/1/96
   -- version 1.1.0D

dss.h
/*
* $Id: dss.h,v 1.10 2008/03/21 18:07:13 jms Exp $
*
* Revision History
* ===================
* $Log: dss.h,v $
* Revision 1.10  2008/03/21 18:07:13  jms
* update copyright date
*
* Revision 1.9  2006/07/31 17:23:09  jms
* fix to parallelism problem
*
* Revision 1.8  2006/03/09 18:55:29  jms
* remove vestigial cvs merge marker
*
* Revision 1.7  2005/10/28 03:05:05  jms
* up maximum scale to 100TB
*
* Revision 1.6  2005/10/28 02:55:26  jms
* add release.h changes
*
* Revision 1.5  2005/10/27 18:13:03  jms
* a_rnd() prototype correction
*
* Revision 1.4  2005/10/25 17:58:59  jms
* update version stamp
*
* Revision 1.3  2005/03/04 19:48:39  jms
* Changes from Doug Johnson to address very large scale factors
*
* Revision 1.2  2005/01/03 20:08:58  jms
* change line terminations
*
* Revision 1.1.1.1  2004/11/24 23:31:46  jms
* re-establish external server
*
* Revision 1.5  2004/04/08 17:34:15  jms
* cleanup SOLARIS/SUN ifdefs; now all use SUN
*
* Revision 1.4  2004/04/07 20:17:29  jms
* bug #58 (join fails between order/lineitem)
*
* Revision 1.3  2004/03/16 14:37:53  jms
* update version and copyright date; correct comment typo
*
* Revision 1.2  2004/02/18 14:07:20  jms
* change to version 2.1.0
*
* Revision 1.1.1.1  2003/08/08 21:50:33  jms
* recreation after CVS crash
*
* Revision 1.3  2003/08/08 21:35:26  jms
* first integration of rng64 for o_custkey and l_partkey
*
* Revision 1.2  2003/08/07 17:58:34  jms
* Convery RNG to 64bit space as preparation for new large scale RNG
*
* Revision 1.1.1.1  2003/04/03 18:54:21  jms
* initial checkin
*
*
*/
 /*
 * general definitions and control information for the DSS code 
 * generator; if it controls the data set, it's here
 */
#ifndef DSS_H
#define  DSS_H
#ifdef TPCH
#define NAME			"TPC-H"
#endif
#ifdef TPCR
#define NAME			"TPC-R"
#endif
#ifndef NAME
#error Benchmark version must be defined in config.h
#endif
#define TPC             "Transaction Processing Performance Council"
#define C_DATES         "1994 - 2010"

#include "config.h"
#include "shared.h"
#include "jcch_dbgen.h"

#include <stdio.h>
#include <stdlib.h>

#define  NONE		-1
#define  PART		0
#define  PSUPP		1
#define  SUPP		2
#define  CUST		3
#define  ORDER		4
#define  LINE		5
#define  ORDER_LINE 6
#define  PART_PSUPP 7
#define  NATION		8
#define  REGION		9
#define  UPDATE		10
#define  MAX_TABLE	11
#define  ONE_STREAM	1
#define  ADD_AT_END	2

#ifdef MAX
#undef MAX
#endif
#ifdef MIN
#undef MIN
#endif
#define MAX(a,b) ((a > b )?a:b)
#define MIN(A,B)  ( (A) < (B) ? (A) : (B))

#define INTERNAL_ERROR(p)  {fprintf(stderr,"%s", p);abort();}
#define LN_CNT  4
static char lnoise[4] = {'|', '/', '-', '\\' };
#define LIFENOISE(n, var)	\
	if (verbose > 0) fprintf(stderr, "%c\b", lnoise[(var%LN_CNT)])

#define MALLOC_CHECK(var) \
    if ((var) == NULL) \
        { \
        fprintf(stderr, "Malloc failed at %s:%d\n",  \
            __FILE__, __LINE__); \
        exit(1);\
        }
#define OPEN_CHECK(var, path) \
    if ((var) == NULL) \
        { \
        fprintf(stderr, "Open failed for %s at %s:%d\n",  \
            path, __FILE__, __LINE__); \
        exit(1);\
        }
#ifndef MAX_CHILDREN
#define MAX_CHILDREN    1000
#endif

/*
 * macros that control sparse keys
 *
 * refer to Porting.Notes for a complete explanation
 */
#ifndef BITS_PER_LONG
#define BITS_PER_LONG   32
#define MAX_LONG        0x7FFFFFFF
#endif /* BITS_PER_LONG */
#define SPARSE_BITS      2
#define SPARSE_KEEP      3
#define  MK_SPARSE(key, seq) \
         (((((key>>3)<<2)|(seq & 0x0003))<<3)|(key & 0x0007))

#define RANDOM(tgt, lower, upper, stream)	dss_random(&tgt, lower, upper, stream)
#define RANDOM64(tgt, lower, upper, stream)	dss_random64(&tgt, lower, upper, stream)
	
     

// typedef struct
// {
//    long      weight;
//    char     *text;
// }         set_member;

// typedef struct
// {
//    int      count;
//    int      max;
//    set_member *list;
//    long *permute;
// }         distribution;
/*
 * some handy access functions 
 */
#define DIST_SIZE(d)		d->count
#define DIST_MEMBER(d, i)	((set_member *)((d)->list + i))->text
#define DIST_PERMUTE(d, i)	(d->permute[i])

typedef struct
{
   char     *name;
   char     *comment;
   DSS_HUGE      base;
   int       (*loader) ();
   long      (*gen_seed)();
   int       child;
   DSS_HUGE vtotal;
}         tdef;

typedef struct SEED_T {
	long table;
	DSS_HUGE value;
	DSS_HUGE usage;
	DSS_HUGE boundary;
#ifdef RNG_TEST
	DSS_HUGE nCalls;
#endif
	} seed_t;


#if defined(__STDC__)
#define PROTO(s) s
#else
#define PROTO(s) ()
#endif

/* bm_utils.c */
char	*env_config PROTO((char *var, char *dflt));
long	yes_no PROTO((char *prompt));
void     a_rnd PROTO((int min, int max, int column, char *dest));
int     tx_rnd PROTO((long min, long max, long column, char *tgt));
long	julian PROTO((long date));
long	unjulian PROTO((long date));
FILE	*tbl_open PROTO((int tbl, char *mode));
long	dssncasecmp PROTO((char *s1, char *s2, int n));
long	dsscasecmp PROTO((char *s1, char *s2));
int		pick_str PROTO((distribution * s, int c, char *target));
void	agg_str PROTO((distribution *set, long count, long col, char *dest));
void	read_dist PROTO((char *path, char *name, distribution * target));
void	embed_str PROTO((distribution *d, int min, int max, int stream, char *dest));
#ifndef STDLIB_HAS_GETOPT
int		getopt PROTO((int arg_cnt, char **arg_vect, char *oprions));
#endif /* STDLIB_HAS_GETOPT */
DSS_HUGE	set_state PROTO((int t, long scale, long procs, long step, DSS_HUGE *e));

/* rnd.c */
DSS_HUGE	NextRand PROTO((DSS_HUGE nSeed));
DSS_HUGE	UnifInt PROTO((DSS_HUGE nLow, DSS_HUGE nHigh, long nStream));
void	dss_random(DSS_HUGE *tgt, DSS_HUGE min, DSS_HUGE max, long seed);
void	row_start(int t);
void	row_stop(int t);
void	dump_seeds(int t);

/* text.c */
#define MAX_GRAMMAR_LEN	12	/* max length of grammar component */
#define MAX_SENT_LEN	256 /* max length of populated sentence */
#define RNG_PER_SENT	27	/* max number of RNG calls per sentence */

void		dbg_text PROTO((char * t, int min, int max, int s));

#ifdef DECLARER
#define EXTERN
#else
#define EXTERN extern
#endif            /* DECLARER */


// EXTERN distribution nations;
// EXTERN distribution nations2;
// EXTERN distribution regions;
// EXTERN distribution o_priority_set;
// EXTERN distribution l_instruct_set;
// EXTERN distribution l_smode_set;
// EXTERN distribution l_category_set;
// EXTERN distribution l_rflag_set;
// EXTERN distribution c_mseg_set;
// EXTERN distribution colors;
// EXTERN distribution p_types_set;
// EXTERN distribution p_cntr_set;

/* distributions that control text generation */
// EXTERN distribution articles;
// EXTERN distribution nouns;
// EXTERN distribution adjectives;
// EXTERN distribution adverbs;
// EXTERN distribution prepositions;
// EXTERN distribution verbs;
// EXTERN distribution terminators;
// EXTERN distribution auxillaries;
// EXTERN distribution np;
// EXTERN distribution vp;
// EXTERN distribution grammar;


EXTERN long scale;
EXTERN int refresh;
EXTERN int resume;
EXTERN long verbose;
EXTERN long force;
EXTERN long updates;
EXTERN long table;
EXTERN long children;
EXTERN int  step;
EXTERN int	set_seeds;
EXTERN char *d_path;

/* added for segmented updates */
EXTERN int insert_segments;
EXTERN int delete_segments;
EXTERN int insert_orders_segment;
EXTERN int insert_lineitem_segment;
EXTERN int delete_segment;
 

#ifndef DECLARER
extern tdef tdefs[];

#endif            /* DECLARER */


/*****************************************************************
 ** table level defines use the following naming convention: t_ccc_xxx
 ** with: t, a table identifier
 **       ccc, a column identifier
 **       xxx, a limit type
 ****************************************************************
 */

/*
 * defines which control the parts table
 */
#define  P_SIZE       126
#define  P_NAME_SCL   5
#define  P_MFG_TAG    "Manufacturer#"
#define  P_MFG_FMT     "%%s%%0%d%s"
#define  P_MFG_MIN     1
#define  P_MFG_MAX     5
#define  P_BRND_TAG   "Brand#"
#define  P_BRND_FMT   "%%s%%0%d%s"
#define  P_BRND_MIN     1
#define  P_BRND_MAX     5
#define  P_SIZE_MIN    1
#define  P_SIZE_MAX    50
#define  P_MCST_MIN    100
#define  P_MCST_MAX    99900
#define  P_MCST_SCL    100.0
#define  P_RCST_MIN    90000
#define  P_RCST_MAX    200000
#define  P_RCST_SCL    100.0
/*
 * defines which control the suppliers table
 */
#define  S_SIZE     145
#define  S_NAME_TAG "Supplier#"
#define  S_NAME_FMT "%%s%%0%d%s"
#define  S_ABAL_MIN   -99999
#define  S_ABAL_MAX    999999
#define  S_CMNT_MAX    101      
#define  S_CMNT_BBB    10       /* number of BBB comments/SF */
#define  BBB_DEADBEATS 50       /* % that are complaints */
#define  BBB_BASE  "Customer "
#define  BBB_COMPLAIN  "Complaints"
#define  BBB_COMMEND   "Recommends"
#define  BBB_CMNT_LEN  19
#define  BBB_BASE_LEN  9
#define  BBB_TYPE_LEN  10

/*
 * defines which control the partsupp table
 */
#define  PS_SIZE      145
#define  PS_SKEY_MIN  0
#define  PS_SKEY_MAX  ((tdefs[SUPP].base - 1) * scale)
#define  PS_SCST_MIN  100
#define  PS_SCST_MAX  100000
#define  PS_QTY_MIN   1
#define  PS_QTY_MAX   9999
/*
 * defines which control the customers table
 */
#define  C_SIZE       165
#define  C_NAME_TAG   "Customer#"
#define  C_NAME_FMT   "%%s%%0%d%s"
#define  C_MSEG_MAX    5
#define  C_ABAL_MIN   -99999
#define  C_ABAL_MAX    999999
/*
 * defines which control the order table
 */
#define  O_SIZE          109
#define  O_CKEY_MIN      1
#define  O_CKEY_MAX      (tdefs[CUST].base * scale)
#define  O_ODATE_MIN     STARTDATE
#define  O_ODATE_MAX     (STARTDATE + TOTDATE - \
                         (L_SDTE_MAX + L_RDTE_MAX) - 1)
#define  O_CLRK_TAG      "Clerk#"
#define  O_CLRK_FMT   "%%s%%0%d%s"
#define  O_CLRK_SCL      1000
#define  O_LCNT_MIN      1
#define  O_LCNT_MAX      7
#ifndef JCCH_SKEW
#define  MAX_L_PER_O     O_LCNT_MAX
#else
#define  MAX_L_PER_O	(tdefs[LINE].base * scale / 5) 
#endif

/*
 * defines which control the lineitem table
 */
#define  L_SIZE       144L
#define  L_QTY_MIN    1
#define  L_QTY_MAX    50
#define  L_TAX_MIN    0
#define  L_TAX_MAX    8
#define  L_DCNT_MIN   0
#define  L_DCNT_MAX   10
#define  L_PKEY_MIN   1
#define  L_PKEY_MAX   (tdefs[PART].base * scale)
#define  L_SDTE_MIN   1
#define  L_SDTE_MAX   121
#define  L_CDTE_MIN   30
#define  L_CDTE_MAX   90
#define  L_RDTE_MIN   1
#define  L_RDTE_MAX   30
/*
 * defines which control the time table
 */
#define  T_SIZE       30
#define  T_START_DAY  3     /* wednesday ? */
#define  LEAP(y)  ((!(y % 4) && (y % 100))?1:0)

/*******************************************************************
 *******************************************************************
 ***
 *** general or inter table defines
 ***
 *******************************************************************
 *******************************************************************/
#define  SUPP_PER_PART 4
#define  ORDERS_PER_CUST 10 /* sync this with CUST_MORTALITY */
#define  CUST_MORTALITY 3  /* portion with have no orders */
#define  NATIONS_MAX  90 /* limited by country codes in phone numbers */
#define  PHONE_FMT    "%02d-%03d-%03d-%04d"
#define  STARTDATE    92001
#define  CURRENTDATE  95168
#define  ENDDATE      98365
#define  TOTDATE      2557
#define  UPD_PCT      10
#define  MAX_STREAM   47
#define  V_STR_LOW    0.4
#define  PENNIES    100 /* for scaled int money arithmetic */
#define  Q11_FRACTION (double)0.0001
/*
 * max and min SF in GB; Larger SF will require changes to the build routines
 */
#define  MIN_SCALE      1.0
#define  MAX_SCALE   100000.0
/*
 * beyond this point we need to allow for BCD calculations
 */
#define  MAX_32B_SCALE   1000.0
#define LONG2HUGE(src, dst)		*dst = (DSS_HUGE)src	
#define HUGE2LONG(src, dst)		*dst = (long)src
#define HUGE_SET(src, dst)		*dst = *src	
#define HUGE_MUL(op1, op2)		*op1 *= op2	
#define HUGE_DIV(op1, op2)		*op1 /= op2	
#define HUGE_ADD(op1, op2, dst)	*dst = *op1 + op2	
#define HUGE_SUB(op1, op2, dst)	*dst = *op1 - op2	
#define HUGE_MOD(op1, op2)		*op1 % op2	
#define HUGE_CMP(op1, op2)		(*op1 == *op2)?0:(*op1 < *op2)-1:1


/******** environmental variables and defaults ***************/
#define  DIST_TAG  "DSS_DIST"		/* environment var to override ... */
#define  DIST_DFLT "dists.dss"		/* default file to hold distributions */
#define  PATH_TAG  "DSS_PATH"		/* environment var to override ... */
#define  PATH_DFLT "."				/* default directory to hold tables */
#define  CONFIG_TAG  "DSS_CONFIG"	/* environment var to override ... */
#define  CONFIG_DFLT "."			/* default directory to config files */
#define  ADHOC_TAG  "DSS_ADHOC"		/* environment var to override ... */
#define  ADHOC_DFLT "adhoc.dss"		/* default file name for adhoc vars */

/******* output macros ********/
#ifndef SEPARATOR
#define SEPARATOR '|' /* field spearator for generated flat files */
#endif
/* Data type flags for a single print routine */
#define DT_STR		0
#ifndef MVS
#define DT_VSTR		DT_STR
#else
#define DT_VSTR		1
#endif /* MVS */
#define DT_INT		2
#define DT_HUGE		3
#define DT_KEY		4
#define DT_MONEY	5
#define DT_CHR		6

int dbg_print(int dt, FILE *tgt, void *data, int len, int eol);
#define PR_STR(f, str, len)		dbg_print(DT_STR, f, (void *)str, len, 1)
#define PR_VSTR(f, str, len) 	dbg_print(DT_VSTR, f, (void *)str, len, 1)
#define PR_VSTR_LAST(f, str, len) 	dbg_print(DT_VSTR, f, (void *)str, len, 0)
#define PR_INT(f, str) 			dbg_print(DT_INT, f, (void *)str, 0, 1)
#define PR_HUGE(f, str) 		dbg_print(DT_HUGE, f, (void *)str, 0, 1)
#define PR_HUGE_LAST(f, str)        dbg_print(DT_HUGE, f, (void *)str, 0, 0)
#define PR_KEY(f, str) 			dbg_print(DT_KEY, f, (void *)str, 0, -1)
#define PR_MONEY(f, str) 		dbg_print(DT_MONEY, f, (void *)str, 0, 1)
#define PR_CHR(f, str)	 		dbg_print(DT_CHR, f, (void *)str, 0, 1)
#define  PR_STRT(fp)   /* any line prep for a record goes here */
#define  PR_END(fp)    fprintf(fp, "\n")   /* finish the record here */
#ifdef MDY_DATE
#define  PR_DATE(tgt, yr, mn, dy)	\
   sprintf(tgt, "%02d-%02d-19%02d", mn, dy, yr)
#else
#define  PR_DATE(tgt, yr, mn, dy)	\
sprintf(tgt, "19%02d-%02d-%02d", yr, mn, dy)
#endif /* DATE_FORMAT */

/*
 * verification macros
 */
#define  VRF_STR(t, d) {char *xx = d; while (*xx) tdefs[t].vtotal += *xx++;}
#define  VRF_INT(t,d)  tdefs[t].vtotal += d
#define  VRF_HUGE(t,d)	tdefs[t].vtotal = *((long *)&d) + *((long *)(&d + 1))
/* assume float is a 64 bit quantity */
#define  VRF_MONEY(t,d)	tdefs[t].vtotal = *((long *)&d) + *((long *)(&d + 1))
#define  VRF_CHR(t,d)	tdefs[t].vtotal += d
#define  VRF_STRT(t)  
#define  VRF_END(t)  

/*********** distribuitons currently defined *************/
#define  UNIFORM   0

/*
 * seed indexes; used to separate the generation of individual columns
 */
#define  P_MFG_SD  0
#define  P_BRND_SD 1
#define  P_TYPE_SD 2
#define  P_SIZE_SD 3
#define  P_CNTR_SD 4
#define  P_RCST_SD 5
#define  PS_QTY_SD 7
#define  PS_SCST_SD   8
#define  O_SUPP_SD 10
#define  O_CLRK_SD 11
#define  O_ODATE_SD   13
#define  L_QTY_SD  14
#define  L_DCNT_SD 15
#define  L_TAX_SD  16
#define  L_SHIP_SD 17
#define  L_SMODE_SD   18
#define  L_PKEY_SD 19
#define  L_SKEY_SD 20
#define  L_SDTE_SD 21
#define  L_CDTE_SD 22
#define  L_RDTE_SD 23
#define  L_RFLG_SD 24
#define  C_NTRG_SD 27
#define  C_PHNE_SD 28
#define  C_ABAL_SD 29
#define  C_MSEG_SD 30
#define  S_NTRG_SD 33
#define  S_PHNE_SD 34
#define  S_ABAL_SD 35
#define  P_NAME_SD 37
#define  O_PRIO_SD 38
#define  HVAR_SD   39
#define  O_CKEY_SD 40
#define  N_CMNT_SD 41
#define  R_CMNT_SD 42
#define  O_LCNT_SD 43
#define  BBB_JNK_SD    44          
#define  BBB_TYPE_SD   45         
#define  BBB_CMNT_SD   46         
#define  BBB_OFFSET_SD 47         
#endif            /* DSS_H */
